freqtrade/exchange/exchange.py

```diff
@@ -1852,26 +1855,119 @@ def _get_funding_fees_from_exchange(self, pair: str, since: Union[datetime, int]
         except ccxt.BaseError as e:
             raise OperationalException(e) from e
 
-    def fill_leverage_brackets(self):
+    def load_leverage_tiers(self) -> Dict[str, List[Dict]]:
+        if self.trading_mode == TradingMode.FUTURES and self._api.has['fetchLeverageTiers']:
+            try:
+                return self._api.fetch_leverage_tiers()
+            except ccxt.DDoSProtection as e:
+                raise DDosProtection(e) from e
+            except (ccxt.NetworkError, ccxt.ExchangeError) as e:
+                raise TemporaryError(
+                    f'Could not load leverage tiers due to {e.__class__.__name__}.'
+                    f'Message: {e}'
+                ) from e
+            except ccxt.BaseError as e:
+                raise OperationalException(e) from e
+        else:
+            return {}
+
+    @retrier
+    def fill_leverage_tiers(self) -> None:
         """
-        Assigns property _leverage_brackets to a dictionary of information about the leverage
+        Assigns property _leverage_tiers to a dictionary of information about the leverage
         allowed on each pair
-        Not used if the exchange has a static max leverage value for the account or each pair
         """
-        return
+        leverage_tiers = self.load_leverage_tiers()
+        for pair, tiers in leverage_tiers.items():
+            pair_tiers = []
+            for tier in tiers:
+                pair_tiers.append(self.parse_leverage_tier(tier))
+            self._leverage_tiers[pair] = pair_tiers
+
+    def parse_leverage_tier(self, tier) -> Dict:
+        info = tier['info']
+        return {
+            'min': tier['notionalFloor'],
+            'max': tier['notionalCap'],
+            'mmr': tier['maintenanceMarginRatio'],
+            'lev': tier['maxLeverage'],
+            'maintAmt': float(info['cum']) if 'cum' in info else None,
+        }
 
     def get_max_leverage(self, pair: str, stake_amount: Optional[float]) -> float:
         """
         Returns the maximum leverage that a pair can be traded at
         :param pair: The base/quote currency pair being traded
-        :param nominal_value: The total value of the trade in quote currency (margin_mode + debt)
+        :stake_amount: The total value of the traders margin_mode in quote currency
         """
-        market = self.markets[pair]
-        if market['limits']['leverage']['max'] is not None:
-            return market['limits']['leverage']['max']
-        else:
+
+        if self.trading_mode == TradingMode.SPOT:
             return 1.0
 
+        if self._api.has['fetchLeverageTiers']:
+
+            # Checks and edge cases
+            if stake_amount is None:
+                raise OperationalException(
+                    'binance.get_max_leverage requires argument stake_amount')
+            if pair not in self._leverage_tiers:
+                tiers = self.get_leverage_tiers_for_pair(pair)
```

Comment:
If you use any "active" calls here (calls that need the ccxt object) - then we cannot use this method in backtesting/hyperopt.

The same will also become a problem for calculating the liquidation price - but for backtesting - it is a hard requirement to first gather all data from the exchange - and only then start backtesting (it's because of hyperopt, as well as for performance reasons - as we might end up calling the same method 1000nds of times for the same pair).


I'm also not sure why this change is necessary - the prior method seemed to contain what's needed (at least for binance). 
The traffic / memory requirement also seems to heavily increase (we're going from 1kb response to a 16kb response (which we will also have to keep in memory) in case of binance for example).

Reply:
> The traffic / memory requirement also seems to heavily increase (we're going from 1kb response to a 16kb response (which we will also have to keep in memory) in case of binance for example).

We don't store the whole response in memory, and the calculation of max leverage was incorrect before, I don't think you can calculate the max leverage from just the maintenance margin ratio, the maintenance amount calculation was wrong too I think.